mono/metadata/threads.c

```diff
@@ -3785,28 +3787,54 @@ dump_thread (MonoInternalThread *thread, ThreadDumpUserData *ud)
 		}
 	}
 
-	fprintf (stdout, "%s", text->str);
+	fprintf (output_file, "%s\n", text->str);
 
 #if PLATFORM_WIN32 && TARGET_WIN32 && _DEBUG
 	OutputDebugStringA(text->str);
 #endif
 
 	g_string_free (text, TRUE);
-	fflush (stdout);
+	fflush (output_file);
 }
 
 void
 mono_threads_perform_thread_dump (void)
 {
+	FILE* output_file = NULL;
 	ThreadDumpUserData ud;
 	MonoInternalThread *thread_array [128];
 	int tindex, nthreads;
 
 	if (!thread_dump_requested)
 		return;
-
-	printf ("Full thread dump:\n");
-
+	if (thread_dump_dir != NULL) {
+		GString* path = g_string_new (0);
+		char time_str[80];
+		struct timeval tv;
+		long ms;
+#ifdef HAVE_LOCALTIME_R
+		struct tm tod;
+		if (gettimeofday (&tv, NULL) == -1) {
+			g_error ("gettimeofday() failed; errno is %d (%s)", errno, strerror (errno));
+		}
+		localtime_r(&tv.tv_sec, &tod);
+		strftime(time_str, sizeof(time_str), "%Y-%m-%d %H:%M:%S", &tod);
+#else
```

Comment:
we have a pretty similar time formatting logic in `mono_log_write_logfile ()`: https://github.com/mono/mono/blob/8d30d41909d7c67cca4051156e0ef8331d0fe536/mono/utils/mono-log-common.c#L110-L120

Should we refactor to a utility function?

Reply:
I see the code in mono-log-common but I need ms resolution in my case. I think it can be helpful utility code.